File: back-end/schemas/document.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel
from pydantic import Field


class DocumentUploadResponse(BaseModel):
    document_id: UUID
    task_id: UUID
    status: str
    ocr_markdown: str | None = None
    embedding_model: str | None = None


class DocumentStatusResponse(BaseModel):
    document_id: UUID
    task_id: UUID | None = None
    status: str
    stage: str | None = None
    progress: int = 0
    message: str | None = None


class DocumentMarkdownResponse(BaseModel):
    document_id: UUID
    status: str
    markdown: str | None = None
    embedding_model: str | None = None


class DocumentSummaryResponse(BaseModel):
    document_id: UUID
    file_name: str
    status: str
    summary: str | None = None
    # documents.keywords에 저장된 문서 대표 키워드를 프론트 요약 화면에 내려준다.
    keywords: list[str] = []
    page_count: int | None = 0
    file_size: int
    upload_at: datetime
    process_at: datetime | None = None
    embedding_model: str | None = None
    llm_model: str | None = None
    category: dict[str, str | float | None] | None = None


class DocumentActionResponse(BaseModel):
    document_id: UUID
    task_id: UUID | None = None
    status: str
    message: str


class DocumentDeleteResponse(BaseModel):
    document_id: UUID
    file_name: str
    deleted: bool
    message: str


class DocumentChatRequest(BaseModel):
    message: str


class DocumentChatCitation(BaseModel):
    source: str
    label: str
    chunk_id: UUID | None = None
    page_no: int | None = None


class DocumentChatResponse(BaseModel):
    answer: str
    citations: list[DocumentChatCitation] = []
    session_id: UUID | None = None
    message_id: UUID | None = None


class DocumentChatSessionCreateRequest(BaseModel):
    title: str | None = None


class DocumentChatSessionListItem(BaseModel):
    id: UUID
    title: str
    message_count: int
    created_at: datetime
    updated_at: datetime


class DocumentChatMessageResponse(BaseModel):
    id: UUID
    role: str
    content: str
    created_at: datetime


class DocumentChatSessionDetailResponse(BaseModel):
    id: UUID
    title: str
    message_count: int
    created_at: datetime
    updated_at: datetime
    messages: list[DocumentChatMessageResponse] = Field(default_factory=list)


class DocumentResponse(BaseModel):
    id: UUID
    file_name: str
    file_size: int
    status: str
    category: str | None = None
    category_confidence: float | None = None
    keywords: list[str] = []
    summary: str | None = None
    page_count: int | None = 0
    selected_embedding_model: str | None = None
    upload_at: datetime

    class Config:
        from_attributes = True

File: front-end/src/types/document.ts
// 문서 자체 상태
export type DocumentStatus =
  | "PENDING"
  | "PROCESSING"
  | "REVIEW_REQUIRED"
  | "COMPLETED"
  | "FAILED";

export type TaskStatus =
  | "PENDING"
  | "PROCESSING"
  | "COMPLETED"
  | "FAILED";

export type TaskType =
  | "OCR"
  | "SUMMARY"
  | "EMBEDDING"
  | "RAG_INDEXING";

export type TaskStage = TaskType;

export type PipelineStepState =
  | "pending"
  | "processing"
  | "completed"
  | "failed";

export type PipelineStepId =
  | "upload"
  | "extraction"
  | TaskStage;

export interface PipelineStep {
  id: PipelineStepId;
  label: string;
  state: PipelineStepState;
}


export interface Document {
  id: string;
  name: string;
  uploadDate: string;
  status: DocumentStatus;
  stage?: TaskStage;
  progress: number;
  pages?: number;
  summary?: string;
}

// DocumentList에서 사용하는 interface
// 추후 개발 시 위에 Document와 합쳐야 함.(지금은 임시)
export interface DocumentList {
  id: string;
  name: string;
  uploadDate: string;
  size: string;
  pages: number;
  status: DocumentStatus;
  stage?: TaskStage;
  category?: string;
  summary?: string;
  progress?: number;
}

export interface DocumentItem {
  id: string;
  file_name: string;
  status: DocumentStatus;
  category: string | null;
  summary?: string | null;
  page_count: number | null;
  selected_embedding_model: string | null;
  upload_at: string;
}

// Dashboard List 상태 화면 표시용
export type DocumentDisplayStatus =
  | "processing"
  | "review-required"
  | "completed"
  | "failed";

export type DocumentProcessingStatus =
  | 'PENDING'
  | 'PROCESSING'
  | 'COMPLETED'
  | 'FAILED'
  | 'SUCCESS'
  | 'FAILURE'
  | 'REVIEW_REQUIRED';

export interface DocumentUploadResponse {
  document_id: string;
  task_id: string;
  status: DocumentProcessingStatus | string;
  embedding_model?: string | null;
}

export interface DocumentStatusResponse {
  document_id: string;
  task_id: string | null;
  status: DocumentProcessingStatus | string;
  stage: TaskStage | string | null;
  progress: number;
  message: string | null;
}

export type DocumentStatusEvent = DocumentStatusResponse;

export interface DocumentApiError {
  detail?: string;
  message?: string;
}

export interface DocumentMarkdownResponse {
  document_id: string;
  status: DocumentProcessingStatus | string;
  markdown: string | null;
  embedding_model: string | null;
}

export interface DocumentSummaryResponse {
  document_id: string;
  file_name: string;
  status: DocumentProcessingStatus | string;
  summary: string | null;
  keywords?: string[];
  page_count: number | null;
  file_size: number;
  upload_at: string;
  process_at: string | null;
  embedding_model: string | null;
  llm_model: string | null;
}

export interface DocumentActionResponse {
  document_id: string;
  task_id: string | null;
  status: DocumentProcessingStatus | string;
  message: string;
}

export interface EmbeddingModelOption {
  value: string;
  label: string;
}

export interface EmbeddingModelsResponse {
  default_model: string;
  models: EmbeddingModelOption[];
}
